End of week 1

ETLanza opened the pull request.

ETLanza committed:
End of week 1

lanza commented on week1/1.5/Notes/Notes/Controllers/Model Controllers/NoteController.swift:
```diff
@@ -0,0 +1,63 @@
+    
+    static let shared = NoteController()
+    
+    var notes = [Note]()
```
var notes: [Note] = []

lanza replied:
var notes: [Note] = []

lanza commented on week1/1.5/Notes/Notes/Controllers/Model Controllers/NoteController.swift:
```diff
@@ -0,0 +1,63 @@
+
+class NoteController: Codable {
+    
+    static let shared = NoteController()
```
This is a debatable technique. There's a lot of anti-global variable opinions in the world, but that's a topic for future learning of design pattern.

lanza commented on week1/1.5/Notes/Notes/Controllers/Model Controllers/NoteController.swift:
```diff
@@ -0,0 +1,63 @@
+    }
+    
+    func deleteNote(_ note: Note) {
+        guard let indexOfNote = NoteController.shared.notes.index(of: note) else { return }
```
This function probably shouldn't fail silently. Not a big issue for a simple little test app but if you were to improve this app then it should return something like `Note?` where it returned the deleted note if found and nil otherwise.

lanza replied:
As a general rule of thumb you can favor shorter variable names if they are created and immediately used. e.g. does the name `indexOfNote` save you from any thinking or searching as compared to the name `index` here? If the variable is going to propagate further through the code than you might ask "`index` of what?" but since you're using it immediately then it's not ambiguous.

lanza commented on week1/1.5/Notes/Notes/Controllers/Model Controllers/NoteController.swift:
```diff
@@ -0,0 +1,63 @@
+        let documentDirectory = paths[0]
+        let fileName = "notes.json"
+        let url = documentDirectory.appendingPathComponent(fileName)
+        return url
```
Simple optimization but you could just cache this.

    static var url: URL?
    if let url = url {
        return url
    } else {
        // compute the path
    }

lanza commented on week1/1.5/Notes/Notes/Controllers/Model Controllers/NoteController.swift:
```diff
@@ -0,0 +1,63 @@
+    }
+    
+    func saveToPersistentStore() {
+        let jsonEncoder = JSONEncoder()
```
static let encoder = JSONEncoder()

is a similar optimization as mentioned above.

lanza commented on week1/1.5/Notes/Notes/Controllers/Model Controllers/NoteController.swift:
```diff
@@ -0,0 +1,63 @@
+            let data = try jsonEncoder.encode(notes)
+            try data.write(to: fileURL())
+        } catch let error {
+            print("Error saving to persistent store: \(error.localizedDescription)")
```
Another issue that's not important for a test app, but what would a user prefer happened here? Imagine you can't save your notes in your notes app because there was an error in the save attempt. It should probably pick a series of backup locations and attempt to save there. Maybe on launch the app could check for any sign of a failed save and try to perform recovery.

lanza commented on week1/1.5/Notes/Notes/Controllers/View Controllers/NoteDetailViewController.swift:
```diff
@@ -0,0 +1,36 @@
+    
+    //MARK: - IBAction
+    @IBAction func saveButtonPressed(_ sender: UIBarButtonItem) {
+        guard let text = bodyTextView.text, text != "" else { return }
```
Another user unfriendly failure. A button shouldn't be active and not responsive. You could disable the button fi there is no text or you could open a quick popup saying "There's not text here" or something.

lanza commented on week1/1.5/Notes/Notes/Controllers/View Controllers/NoteDetailViewController.swift:
```diff
@@ -0,0 +1,36 @@
+    //MARK: - IBAction
+    @IBAction func saveButtonPressed(_ sender: UIBarButtonItem) {
+        guard let text = bodyTextView.text, text != "" else { return }
+        if let note = self.note {
```
if let note = note {

lanza commented on week1/1.5/Notes/Notes/Controllers/View Controllers/NoteDetailViewController.swift:
```diff
@@ -0,0 +1,36 @@
+        if let note = self.note {
+            NoteController.shared.updateNote(note, withBodyText: text)
+        } else {
+            NoteController.shared.addNote(Note.init(bodyText: text))
```
Note(bodyText: text)

lanza commented on week1/1.5/Notes/Notes/Controllers/View Controllers/NoteDetailViewController.swift:
```diff
@@ -0,0 +1,36 @@
+        } else {
+            NoteController.shared.addNote(Note.init(bodyText: text))
+        }
+        let _ = navigationController?.popViewController(animated: true)
```
I forget but I think you can just do `_ = navigation...`

lanza commented on week1/1.5/Notes/Notes/Controllers/View Controllers/NoteDetailViewController.swift:
```diff
@@ -0,0 +1,36 @@
+    
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        if note != nil {
```
if let note = note { 
        bodyTextView.text = note.text
    }

lanza commented on week1/1.5/Notes/Notes/Controllers/View Controllers/NoteTableViewController.swift:
```diff
@@ -0,0 +1,136 @@
+    //MARK: - Properties
+    let searchController = UISearchController(searchResultsController: nil)
+    
+    var filteredNotes = [Note]()
```
var filteredNotes: [Note] = []

lanza commented on week1/1.5/Notes/Notes/Controllers/View Controllers/NoteTableViewController.swift:
```diff
@@ -0,0 +1,136 @@
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        NoteController.shared.loadFromPersistentStore()
+        searchController.searchResultsUpdater = self
```
This code is kinda clustered. Structure it by white space or create a `setupSearchController` method or something.

lanza commented on week1/1.5/Notes/Notes/Controllers/View Controllers/NoteTableViewController.swift:
```diff
@@ -0,0 +1,136 @@
+        if isFiltering() {
+            return filteredNotes.count
+        }
+        return NoteController.shared.notes.count
```
Style preference here but I'm a fan of putting this in the `else` branch.

lanza commented on week1/1.5/Notes/Notes/Controllers/View Controllers/NoteTableViewController.swift:
```diff
@@ -0,0 +1,136 @@
+    
+    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        let cell = tableView.dequeueReusableCell(withIdentifier: "noteCell", for: indexPath)
+        let note = findCorrectNote(withIndex: indexPath)
```
What does `correct` mean? Weird naming.

lanza commented on week1/1.5/Notes/Notes/Controllers/View Controllers/NoteTableViewController.swift:
```diff
@@ -0,0 +1,136 @@
+                let note = findCorrectNote(withIndex: indexPath)
+                NoteController.shared.deleteNote(note)
+                tableView.reloadData()
+                tableView.deleteRows(at: [indexPath], with: .fade)
```
I'm not 100% sure because I don't do iOS anymore but wouldn't `reloadData()` update the table view to no longer have the rows you're about to delete in the next line? I don't recall the subtleties here.

lanza commented on week1/1.5/Notes/Notes/Controllers/View Controllers/NoteTableViewController.swift:
```diff
@@ -0,0 +1,136 @@
+    }
+    
+    // MARK: - Private instance methods
+    func searchBarIsEmpty() -> Bool {
```
Swift has "computed properties" for this purpose.

    var searchBarIsEmpty: Bool { return searchController.searchBar.text?.isEmpty ?? true }

If you prefer to use a function you'd want to name it as one. Something like `checkIfSearchBarIsEmpty()`.

lanza commented on week1/1.5/Notes/Notes/Controllers/View Controllers/NoteTableViewController.swift:
```diff
@@ -0,0 +1,136 @@
+    }
+    
+    func filterContentForSearchText(_ searchText: String, scope: String = "All") {
+        filteredNotes = NoteController.shared.notes.filter({( note : Note) -> Bool in
```
If you know what it is, use trailing closure syntax
    
    sort(items) { a, b in a > b }

instead of
    
    sort(items, { a, b in a < b })

lanza replied:
There should be no space between the `note` and the `:`
    
    (note: Note)

lanza replied:
That being said, you don't need the type declarations here. Anybody reading Swift code generally understands that a filter method call is going to be passed a filter that takes the item and a closure deciding whether or not to filter it.

    NoteController.shared.notes.filter() { return $0.bodyText.lowercased().contains(searchText.lowercased()) }

lanza commented on week1/1.5/Notes/Notes/Controllers/View Controllers/NoteTableViewController.swift:
```diff
@@ -0,0 +1,136 @@
+            editBarButton.title = "Edit";
+            editBarButton.style = UIBarButtonItemStyle.plain;
+        }
+        return searchController.isActive && !searchBarIsEmpty()
```
This isn't too bad, but ugly combinations of multiple different `not`s is ugly. The emptiness is a negative condition. The `!` is a negative condition. This ends up being "active and not not occupied."

lanza commented on week1/1.5/Notes/Notes/Controllers/View Controllers/NoteTableViewController.swift:
```diff
@@ -0,0 +1,136 @@
+        return searchController.isActive && !searchBarIsEmpty()
+    }
+    
+    func findCorrectNote(withIndex index: IndexPath) -> Note {
```
`Correct` is weird. It's an undefined concept here. Do you mean the note corresponding to the index? If so, that's already clear by passing an index.

Also, `withIndex index: IndexPath)` is redundant. `findNote(with indexPath: IndexPath) -> Note`.

lanza commented on week1/1.5/Notes/Notes/Controllers/View Controllers/NoteTableViewController.swift:
```diff
@@ -0,0 +1,136 @@
+    
+    // MARK: - Navigation
+    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
+        guard let destinationVC = segue.destination as? NoteDetailViewController else { return }
```
If this fails the user is confused as hell due too `return`.

lanza commented on week1/1.5/Notes/Notes/Models/Note.swift:
```diff
@@ -0,0 +1,23 @@
+class Note: Equatable, Codable {
+   
+    //MARK: - Properties
+    var bodyText: String
```
`bodyText` seems like a name inspired by the `UIView` element that represents it and not one that refers to the `Note` in and of itself. `body` as opposed to what? -- nothing because there is nothing providing context to explain what would be non-body.

lanza commented on week1/1.5/Notes/Notes/Models/Note.swift:
```diff
@@ -0,0 +1,23 @@
+
+import Foundation
+
+class Note: Equatable, Codable {
```
Have you learned `struct`s yet? This is very structish.

ETLanza committed:
Add source code MARKs

Add MARKs to Life Cycle Methods and moved properties MARK to differentiate the two
X≈

ETLanza committed:
Add Todo project

Simple project to introduce table view data source editing methods
